pkg/kafka/auth/scram.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	TLS "crypto/tls"
+	"hash"
+
+	"github.com/Shopify/sarama"
+	"github.com/pkg/errors"
+	SCRAM "github.com/xdg/scram"
+)
+
+// SCRAMConfig describes the configuration properties required for the SCRAM handshake
+// between the collector and the pipeline
+type SCRAMConfig struct {
+	userName           string
+	password           string
+	algorithm          string
+	useTLS             bool
+	insecureSkipVerify bool
+}
+
+func createTLSConfiguration(b bool) (t *TLS.Config) {
+	t = &TLS.Config{
+		InsecureSkipVerify: b,
+	}
+	return t
+}
+
+func setSCRAMConfiguration(config *SCRAMConfig, saramaConfig *sarama.Config) error {
+	saramaConfig.Net.SASL.Enable = true
+	saramaConfig.Net.SASL.User = config.userName
+	saramaConfig.Net.SASL.Password = config.password
+
+	if config.algorithm == "sha512" {
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA512} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLMechanism(sarama.SASLTypeSCRAMSHA512)
+	} else if config.algorithm == "sha256" {
+		saramaConfig.Net.SASL.SCRAMClientGeneratorFunc = func() sarama.SCRAMClient { return &XDGSCRAMClient{HashGeneratorFcn: SHA256} }
+		saramaConfig.Net.SASL.Mechanism = sarama.SASLMechanism(sarama.SASLTypeSCRAMSHA256)
+	} else {
+		errors.Errorf("invalid SHA algorithm \"%s\": can be either \"sha256\" or \"sha512\"", config.algorithm)
+	}
+
+	if config.useTLS {
+		saramaConfig.Net.TLS.Config = createTLSConfiguration(config.insecureSkipVerify || false)
+		saramaConfig.Net.TLS.Enable = true
+	}
+
+	return nil
+}
+
+// XDGSCRAMClient takes a dereference to a scram client/conversation and a hash function
+type XDGSCRAMClient struct {
```

Comment:
* why XDG prefix?
* does not seem like this struct needs to be public

Reply:
It does need to be exported as it's needed for `pkg/kafka/auth/config.go` file. Unless I'm entirely wrong.